### PR TITLE
Add validator for checking previous period on transaction closure

### PR DESCRIPTION
**Pull request by mreis-ch**
Create validator PreviousPeriodTnClosureValidator to validate previous period exists on small full account only company is multi year filer.
Update CurrentPeriodTnClosureValidator to take the SmallFull object and Errors object. SmallFull object lookup added here as will be used by multiple transaction closure validators rather than each validator call the find. Errors object added as this will be added to by multiple validators
Update AccountsValidator to call new PreviousPeriodTnClosureValidator

BI-5134

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/AccountsValidator.java [hunk 1]
```diff
@@ -1,27 +1,65 @@
...
 
-import javax.servlet.http.HttpServletRequest;
-
 @Component
-public class AccountsValidator {
+public class AccountsValidator extends BaseValidator {
+
+    private static final String SMALL_FULL_PATH = "$company_accounts.small_full";
+
+    private CompanyAccountService companyAccountService;
```
Comment (ch-tstrickland) [inline src/main/java/uk/gov/companieshouse/api/accounts/validation/AccountsValidator.java:24]: These fields could be final

Comment (mreis-ch) [reply src/main/java/uk/gov/companieshouse/api/accounts/validation/AccountsValidator.java:24]: Ok 👍

**Review by faisal-amin82 — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/AccountsValidator.java [hunk 2]
```diff
@@ -1,27 +1,65 @@
...
+            SmallFullService smallFullService, CurrentPeriodTnClosureValidator currentPeriodTnClosureValidator,
+            PreviousPeriodTnClosureValidator previousPeriodTnClosureValidator) {
+        super(companyService);
+        this.companyAccountService = companyAccountService;
+        this.smallFullService = smallFullService;
+        this.currentPeriodTnClosureValidator = currentPeriodTnClosureValidator;
+        this.previousPeriodTnClosureValidator = previousPeriodTnClosureValidator;
+    }
+    
     public Errors validationSubmission(Transaction transaction, String companyAccountsId, HttpServletRequest request)
```
Comment (ch-tstrickland) [inline src/main/java/uk/gov/companieshouse/api/accounts/validation/AccountsValidator.java:43]: validationSubmission isn't a verb (i.e. normal method name). Is this just local convention?

Comment (ch-tstrickland) [reply src/main/java/uk/gov/companieshouse/api/accounts/validation/AccountsValidator.java:43]: ... and yes, I know that this is preexisting code.

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/AccountsValidator.java (hunk 1 above)
Comment (amoriartyCH) [reply src/main/java/uk/gov/companieshouse/api/accounts/validation/AccountsValidator.java]: Just a thought, should the path be "$.company_accounts.small_full" or should it be what it currently is? "$company_accounts.small_full"

Genuine question :)

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/transactionclosure/CurrentPeriodTnClosureValidator.java [hunk 3]
```diff
@@ -1,73 +1,51 @@
...
-                                           SmallFullService smallFullService,
                                            CurrentPeriodService currentPeriodService) {
         super(companyService);
-        this.companyAccountService = companyAccountService;
-        this.smallFullService = smallFullService;
         this.currentPeriodService = currentPeriodService;
     }
 
-    public Errors isValid(String companyAccountsId, HttpServletRequest request) throws DataException {
+    public Errors isValid(String companyAccountsId, SmallFull smallFull, HttpServletRequest request,
```
Comment (ch-tstrickland) [inline src/main/java/uk/gov/companieshouse/api/accounts/validation/transactionclosure/CurrentPeriodTnClosureValidator.java:32]: Sorry to harp on about method names. But...
isFoo-type methods are by convention boolean. This isn't. Maybe change to validate?

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/transactionclosure/PreviousPeriodTnClosureValidator.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import uk.gov.companieshouse.api.accounts.service.response.ResponseStatus;
+import uk.gov.companieshouse.api.accounts.validation.BaseValidator;
+import uk.gov.companieshouse.api.model.transaction.Transaction;
+
+@Component
+public class PreviousPeriodTnClosureValidator extends BaseValidator {
+
+    private final PreviousPeriodService previousPeriodService;
+
+    private static final String SMALL_FULL_PREVIOUS_PERIOD_PATH = "$.small_full.previous_period";
```
Comment (ch-tstrickland) [inline src/main/java/uk/gov/companieshouse/api/accounts/validation/transactionclosure/PreviousPeriodTnClosureValidator.java]: Move statics to top

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/transactionclosure/CurrentPeriodTnClosureValidator.java (hunk 3 above)
Comment (mreis-ch) [reply src/main/java/uk/gov/companieshouse/api/accounts/validation/transactionclosure/CurrentPeriodTnClosureValidator.java]: Done

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/transactionclosure/PreviousPeriodTnClosureValidator.java (hunk 4 above)
Comment (mreis-ch) [reply src/main/java/uk/gov/companieshouse/api/accounts/validation/transactionclosure/PreviousPeriodTnClosureValidator.java]: Done

**Review by ch-tstrickland — APPROVED** [review]

**Pull request merged**